Add possibility to specify own labeldefining commands

Stainzl opened the pull request:
Tried to add feature to set own commands which specify labels for referencing in document (see #813).

Cherrypicked an commit of #803 for testing because I use the same constelation as in this case.

#### Additions
Adds a table to the settings where you can specify commands setting labels and the position of the labelname.

#### Changes
Changed getting labels for inspection and autocompletion

#### Backwards incompatible changes
Hopefully none, I did not delete any old function, only changed the mechanics for such methods

#### To do

As commented [here](https://github.com/Ruben-Sten/TeXiFy-IDEA/pull/815#pullrequestreview-198554223) and [here](https://github.com/Ruben-Sten/TeXiFy-IDEA/pull/815#issuecomment-465515757): 
- [x] Label definitions do not show up in the structure view, neither do `\bibitem`s. Expected behaviour is that all label defining commands show up in the structure view (see `LatexStructureViewElement`).
- [x] The usage of `\bibitem` seems broken as it is not configured by default (using `\cite` and variations).
- [x] Duplicate label inspection is broken.
- [x] Missing label inspection is broken.
- [x] Label convention inspection is broken.
- [x] Make sure to follow the JetBrains [UI guidelines](https://jetbrains.github.io/ui/controls/table/#how-to-use) for the settings table.
- [x] (moved todo from code) rework layout and add info about only required parameters count
- [x] list correct label in goto
- [x] add column to enable command labeling a section or things like this
- [x] Add intention on \newcommand to add a labeldefining command

**Edit**
Included #818 cause it was required for displaying the labels of new defined commands

Stainzl commented:
It would be possible to add the \label command hardcoded but I thought it would be better to make it fully configurable. The \label command should be added to this list by default, so when you do not configure anything the label command work as excepted. If it is not added by default, I have to fix my code.

PHPirates commented:
> Duplicate label inspection is broken.
Missing label inspection is broken.
Label convention inspection is broken.

That's funny, they work for me.

RubenSchellekens commented:
> That's funny, they work for me.

### Settings
![image](https://user-images.githubusercontent.com/17410729/52053495-c4779880-2559-11e9-89c5-b0ca0cbd00a8.png)

### LaTeX
![image](https://user-images.githubusercontent.com/17410729/52053510-d22d1e00-2559-11e9-960c-5d1c95655231.png)

![image](https://user-images.githubusercontent.com/17410729/52053536-e1ac6700-2559-11e9-9a9b-70d503f2c519.png)

Stainzl commented:
Thanks for your response, Ruben. 

- Okay, I didn't notice about the structure view, I will add this.
- Where should the label of \bibitem be used? Because I think, \bibitem isn't specifing a label for \ref style commands or am I wrong? When \bibitem is for \cite style commands only, I would say it has to be added to the functions which look for the bibtex labels.
- You're right, if you define two \label with the same name it gets marked correctly, but when you use another command it will not be marked.
- Missing label has to be done
- Label convention also
- Thanks for the link. I searched for such guidelines, but I think I used the wrong keywords.

RubenSchellekens commented:
> * Where should the label of \bibitem be used? Because I think, \bibitem isn't specifing a label for \ref style commands or am I wrong? When \bibitem is for \cite style commands only, I would say it has to be added to the functions which look for the bibtex labels.

Oddily enough they are extremely intertwined in our current code. If you can separate them, sure go ahead. As long as defined `\bibitem`s show up in the `\cite` (and variations) auto completion.

Stainzl commented:
Okay, this little fix is now included. But I think it isn't shown in structure view by now.

PHPirates commented:
@Stainzl I have copied the review points to the first post so you can check them off when you fix them, and I can see when it is time to review again :) 

> I tried to find the issue with the labels in the structure view, but they are not visible without my changes to. 

I didn't check this but this could be possible of course, that it got broken earlier. But it would be great if you could fix this for this PR, because it would have to be fixed for code added this PR anyway.

Stainzl commented:
> - Missing label inspection is broken.

But it isn't completly broken. When I start a new \section the inspaction correclty highlights that there is no label. After inserting a label the highlighting is removed correctly.
But when an other command comes behind the section command the message is not removed.
The label uses the output of \@currentlabel as seen in this [tex.SO](https://tex.stackexchange.com/a/111281). Because of this I added the label to my new defined command.
Because of this, see in the [MWE in issue](https://github.com/Ruben-Sten/TeXiFy-IDEA/issues/813#issue-403723397), my \listingcaption first defines a caption an specifies the label for this caption. So this command could not be used to label \section or similar commands.
If I understood you correctly you would like to define a command to label a section?
And therefore I need a possibility to decide if the command could label a section or not.

> - Label convention inspection is broken.

When a new defined command could label a section you're right. But when the command itself specifies the counter to label the plugin does not know what the correct convention is in this case, so my idea was to add a field to the command in the table to specify the prefix for this specific command

Stainzl commented:
Okay, the table now nearly fits the guidelines. Does anyone know who to achive the padding in the table header? Because the way I try it actually ignores the border...

PHPirates commented:
Great, a huge improvement!

I am not sure what padding exactly you mean? (I just added one space)
![deepinscreenshot_select-area_20190211162928](https://user-images.githubusercontent.com/15669080/52573518-a09a3980-2e1a-11e9-8d84-356fc337cd6f.png)

Stainzl commented:
Okay, on your picture it looks really nice, but when I start it, it looks like that:
 ![broken table](https://user-images.githubusercontent.com/17581244/52624909-20261800-2eb0-11e9-9e27-396a354ee3a0.png)
I did an `gradlw clean` and run it again, but the table is still broken...

PHPirates commented:
Strange, probably the way I changed the preferred size was not correct. Anyway I see you have made changes, is that fixed now for you? I think the new table height is fine now.

Stainzl commented:
Yes, it is fixed, i changed it completly. I cloned IntelliJ-CE and got inspiration from the git plugin. The table resizes now like the table under `VCS | Git | Remotes...`. The window for adding and editing commands is also inspired by this plugin.
So the table will have show all commands and will have space for three empty lines. Horizontally it will resize to the length of the largest command, so the command will be readable every time.

Stainzl commented on src/nl/rubensten/texifyidea/settings/labelDefiningCommands/MyIntFilter.kt:
```diff
@@ -0,0 +1,60 @@
+            super.insertString(fb, offset, string, attr)
+        }
+        else {
+            TODO("Weiß der Geier was da hin soll...")
```
Yup, I noticed them right after pushing and removed them, because you don't really need them.
This class is used to enable only positiv integer input for the position. I added a comment and updated the class name

Stainzl commented on src/nl/rubensten/texifyidea/settings/labelDefiningCommands/MySettingsTable.kt:
```diff
@@ -0,0 +1,39 @@
+/**
+ * JBTable is extended to set own text when table is empty and define action
+ */
+class MySettingsTable : JBTable {
```
changed name

Stainzl commented on src/nl/rubensten/texifyidea/settings/labelDefiningCommands/MyTableModel.kt:
```diff
@@ -0,0 +1,10 @@
+import com.intellij.util.ui.ItemRemovable
+import javax.swing.table.DefaultTableModel
+
+class MyTableModel : DefaultTableModel(), ItemRemovable {
```
name changed

Stainzl commented on src/nl/rubensten/texifyidea/settings/labelDefiningCommands/TexifyConfigurableLabelCommands.kt:
```diff
@@ -0,0 +1,168 @@
+import javax.swing.table.DefaultTableCellRenderer
+import javax.swing.table.TableCellRenderer
+
+class TexifyConfigurableLabelCommands(private val settings: TexifySettings) {
```
commented whole class

Stainzl commented on src/nl/rubensten/texifyidea/settings/labelDefiningCommands/TexifyDefineLabelingCommand.kt:
```diff
@@ -0,0 +1,70 @@
+import javax.swing.*
+import javax.swing.text.PlainDocument
+
+class TexifyDefineLabelingCommand(cmdName: String, position: Int) : DialogWrapper(null, false) {
```
yep

Stainzl commented:
For the Problem with the table header I need your help, because I don't know really how to reproduce.
I think you added a short command and edited it to a longer command, am I right?
In this case the table wasn't resizing, this is now fixed.
Did I say how much I like developing front end? I love to work with swing ❤️. No, seriously, the label should be under the table now...

PHPirates commented:
:smile: Yes I'm happy that you want to make this into a proper fix. We are really close now I think.

Great, thanks for adding the comments! 

>  I don't know really how to reproduce.

Can you try to run the Gradle clean task and then try again? Because it happens for me on startup, without adding anything.

Stainzl commented:
I tried to reproduce, but even after a gradle clean I do not have the problem...

Stainzl commented:
Yup, but I see in your picture that the command is also shortened. Could you please try to edit a command and make it longer to see if it isn't shortened by now?
When you add a very long command, there will be a horizontal scrollbar, should I try to prevent this?

PHPirates commented:
Yes, that works. No it's fine this way, if anything we should not encourage commands that are so long :smile:

Stainzl commented:
@RubenSchellekens just a little ping because I'm waiting since two weeks for an answer to this [questions](https://github.com/Ruben-Sten/TeXiFy-IDEA/pull/815#issuecomment-459708994).

RubenSchellekens commented:
> I love to work with swing ❤️

Haha, we're lucky to have you then! I hate it :)

The new table looks awesome! Works well, looks nice. Maybe it's an idea to always force a backslash so you won't end up with non-commands (that still get treated as commands)?

> If I understood you correctly you would like to define a command to label a section?
And therefore I need a possibility to decide if the command could label a section or not.

A section should have a label that defines a reference point for that section according to the inspection. That also means that if a section is followed by a custom defined label defining command, the inspection must not trigger. Which it current does. If it works differently, please correct me. Ironically I don't really know much about LaTeX.

> so my idea was to add a field to the command in the table to specify the prefix for this specific command

Sounds like a good idea. When left empty/disabled, the inspection should not be enforced.

And finally I found a little bug with the latest Goto Symbol functionality (#789, #771, #829), it uses the wrong icon and does not take the argument location into account. Should be a simple fix.

![image](https://user-images.githubusercontent.com/17410729/53084964-2ea0af00-3502-11e9-9ec4-ee7fc83ce6fa.png)

RubenSchellekens commented:
I just thought of another enhancement: Make an intention on `\newcommand` like commands that register that command to the table (ideally determining the argument position automagically). This removes some of the hassle of having to navigate through the settings.

Stainzl commented:
> The new table looks awesome! Works well, looks nice. Maybe it's an idea to always force a backslash so you won't end up with non-commands (that still get treated as commands)?

Thanks. The leading backslash is added internally

> Sounds like a good idea. When left empty/disabled, the inspection should not be enforced.

Okay, than I will add this option to the settings.

> I just thought of another enhancement: Make an intention on `\newcommand` like commands that register that command to the table (ideally determining the argument position automagically). This removes some of the hassle of having to navigate through the settings.

Okay, this is a really nice idea, but I would like to add this feature later, if it is okay.

RubenSchellekens commented:
> The leading backslash is added internally

Yeah I noticed. However, I personally think it will look nicer if all commands in the table have a leading backslash.

> Okay, this is a really nice idea, but I would like to add this feature later, if it is okay.

Sure. I'd still include it in this PR though. No need to hurry.

Stainzl commented:
Could you please check that any command you want to define has an leading slash? And that you cant delete or replace it?
(You have to delete any command without a leading slash, because you should not be able to replace the first character of the input.)

PHPirates commented:
I can confirm that I cannot remove or change the leading slash, I like that it is shown so that it is clear it will always be there. Settings are stored between restarts for me.

Stainzl commented:
Okay, at a short look at the stacktrace I think this one is not from this PR, because I haven't changed any of these files in the last days :) But please look for it, and if you find anything I would like ton know it so I can change it.

Stainzl commented:
@RubenSchellekens I added the intention to add the command, but in the way I do it by now it is not liked by IntelliJ. You will get an throwable caused by the add command window. I wont add the command directly, but I have no other idea how to achieve this.

Stainzl commented:
To determine if an new command could label any other command or increases an counter internally I added a hashset of counter increasing commands. I added `\caption`, `\captionof` and the keys of the `labeled` field. Do you know other commands which increase an counter?

PHPirates commented:
> Do you know other commands which increase an counter?

Any user defined command that does so :smile: no but seriously, if those two are enough for you then I'd say that is fine, I think this is not such a critical issue that we need to discuss that lengthily.

I didn't realise until I read your edit that the intention only works with the cursor on the second parameter of \newcommand, if that can be fixed easily that would be great, otherwise I'd leave it as it is. The intention works for me, nice.

You were right btw, I just happened to find a crash coincidentally when working on this branch.

Stainzl commented:
> Any user defined command that does so 😄 no but seriously, if those two are enough for you then I'd say that is fine, I think this is not such a critical issue that we need to discuss that lengthily.

Okay, I thought I could ask, because if anyone has a command in daily use I could add it right now. But anyone who defines himself such a command is able to remove this tick I think 😄 

> I didn't realise until I read your edit that the intention only works with the cursor on the second parameter of \newcommand, if that can be fixed easily that would be great, otherwise I'd leave it as it is. The intention works for me, nice.

Now the intention should work on the whole \newcommand area

> You were right btw, I just happened to find a crash coincidentally when working on this branch.

Okay, lucky me 😄 

I also protected the \label command from beeing removed or edited.

PHPirates commented:
> Now the intention should work on the whole \newcommand area

That works, much better.

> I also protected the \label command from beeing removed or edited.

I just managed to delete it. I think I selected multiple commands and then hit delete. But it's  not important.

Is this PR feature complete now? Or are there more things?

Stainzl commented:
> I just managed to delete it. I think I selected multiple commands and then hit delete. But it's not important.

Okay, I tried my self but was not able to 😄

> Is this PR feature complete now? Or are there more things?

I think so. @RubenSchellekens do you miss anything what should be added?

Stainzl commented:
@RubenSchellekens did I miss something what I should do or are you just to busy to comment here?

PHPirates commented:
@Stainzl Don't worry, I know he'll get to it eventually :) in the meantime I will be including this in alpha releases.

RubenSchellekens commented on src/nl/rubensten/texifyidea/inspections/latex/LatexLabelConventionInspection.kt:
```diff
@@ -67,7 +69,9 @@ open class LatexLabelConventionInspection : TexifyInspectionBase() {
             val prefix = Magic.Command.labeled[context.name]!!
-            if (!required[0].startsWith("$prefix:")) {
+            val position = labelAnyCommands[cmd.name]?.position ?: continue
+            val label = required[position - 1]
```
Crashes when the label is defined with index > 0, but fewer parameters are specified.

```tex
\newcommand{\secondArgumentIsLabel}[2]{Dummy \textbf{#1}\label{#2}}
\section{Broodje}\secondArgumentIsLabel{seca:broodje}
```

Stainzl replied:
fixed in commit ccc748569e2c106beca1ef217771924584d3d3a3

RubenSchellekens commented on src/nl/rubensten/texifyidea/intentions/LatexLabelDefiningNewCommand.kt:
```diff
@@ -0,0 +1,112 @@
+import nl.rubensten.texifyidea.util.*
+
+open class LatexLabelDefiningNewCommand : TexifyIntentionBase("Add label defining command to list") {
+    private val settings = TexifySettings.getInstance()
```
```suggestion

    private val settings = TexifySettings.getInstance()
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/intentions/LatexLabelDefiningNewCommand.kt:
```diff
@@ -0,0 +1,112 @@
+import nl.rubensten.texifyidea.settings.labeldefiningcommands.EditLabelDefiningCommand
+import nl.rubensten.texifyidea.util.*
+
+open class LatexLabelDefiningNewCommand : TexifyIntentionBase("Add label defining command to list") {
```
I'd suggest changing 'list' to 'settings'

```suggestion
open class LatexLabelDefiningNewCommand : TexifyIntentionBase("Add label defining command to settings") {
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/intentions/LatexLabelDefiningNewCommand.kt:
```diff
@@ -0,0 +1,112 @@
+        // get parent element
+        val parentElement = selected.parent
+        val parent = parentElement as? LatexCommands
+                ?: parentElement.parentOfType(LatexCommands::class)
```
can be a oneliner.

Stainzl replied:
fixed in commit 4fc09f4f03cfda54b13164fb3b1d6e31c694f878

Stainzl replied:
fixed in commit 4fc09f4f03cfda54b13164fb3b1d6e31c694f878

RubenSchellekens commented on src/nl/rubensten/texifyidea/intentions/LatexLabelDefiningNewCommand.kt:
```diff
@@ -0,0 +1,112 @@
+        // get parent element
+        val parentElement = selected.parent
+        val parent = parentElement as? LatexCommands
+                ?: parentElement.parentOfType(LatexCommands::class)
```
Can be a oneliner.

Stainzl replied:
fixed in commit 4fc09f4f03cfda54b13164fb3b1d6e31c694f878

RubenSchellekens commented on src/nl/rubensten/texifyidea/intentions/LatexLabelDefiningNewCommand.kt:
```diff
@@ -0,0 +1,112 @@
+        // get the position of the label in new command, identified by the argument number of the first required parameter
+        // of label command
+        val position = label.requiredParameter(0)?.replace("#", "")
+                ?.toIntOrNull() ?: 1
```
Either split it up (each statement on a new line), or put it on one line.

Stainzl replied:
fixed in commit e43885c396a6aec2c17f9d63fb39fa2ae1c7c550

RubenSchellekens commented on src/nl/rubensten/texifyidea/navigation/NavigationItemUtil.kt:
```diff
@@ -17,18 +19,25 @@ object NavigationItemUtil {
+                        text,
+                        if (psiElement.name in Magic.Command.labels ||
+                                TexifySettings.getInstance().labelCommands.containsKey(psiElement.name))
+                            TexifyIcons.DOT_LABEL else TexifyIcons.DOT_BIB
```
Add braces to make it better readable.

```kotlin
                        if (psiElement.name in Magic.Command.labels ||
                                TexifySettings.getInstance().labelCommands.containsKey(psiElement.name)) {
                            TexifyIcons.DOT_LABEL
                        }
                        else TexifyIcons.DOT_BIB
```

Stainzl replied:
Okay, I'm a little bit confused.
```kotlin
if (condition) {
    true
}
else false
```
If I get you right, the `true`-part should be in braces, the `false`-part could be written without braces in case there is only one statement? Because here you added the braces and in this other [comment](https://github.com/Ruben-Sten/TeXiFy-IDEA/pull/815#discussion_r275389152) you removed the braces.

RubenSchellekens replied:
I do that with the `else` clause, when the return value of an if/else statement. It removes a slight bit of clutter in my opinion.

```kotlin
val oneLiner = if (true) 1 else 0

val assignment = if (true) {
    more.complex.stuff().toIntOrNull()
}
else null

if (true) {
    action1()
    action2()
}
else {
    action3()
}
```

Stainzl replied:
Oh, okay, now I understood your schematics. Then I will do it this way.
But I'm a little bit confused again... In this case it was like your `oneLiner` but with a big condition. Should I then use braces true?

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,23 @@
+package nl.rubensten.texifyidea.settings
+
+data class LabelingCommandInformation(var commandName: String, var position: Int, var labelPrevCmd: Boolean) {
+    companion object {
```
```suggestion

    companion object {
    
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,23 @@
+package nl.rubensten.texifyidea.settings
+
+data class LabelingCommandInformation(var commandName: String, var position: Int, var labelPrevCmd: Boolean) {
```
`labelPrevCmd` -> `labelsPreviousCommand`.

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,23 @@
+            val parts = string.split(";")
+            if (parts.size != 3) {
+                throw IllegalArgumentException("String must contain exactly three parts")
+            }
```
```kotlin
require(parts.size == 3) { "String must contain exactly three parts, got ${parts.size}" }
```

Kotlin has these nice contract functions :) Also add a 'got' part for easier debugging.

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,23 @@
+            if (parts.size != 3) {
+                throw IllegalArgumentException("String must contain exactly three parts")
+            }
+            val position = parts[1].toIntOrNull() ?: throw IllegalArgumentException("Second part must be type Int")
```
Not wrong in the slightest, but more Kotlin idiom:
```suggestion
            val position = parts[1].toIntOrNull() ?: error("Second part must be type Int")
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,23 @@
+                throw IllegalArgumentException("String must contain exactly three parts")
+            }
+            val position = parts[1].toIntOrNull() ?: throw IllegalArgumentException("Second part must be type Int")
+            val labelPrevCmd = parts[2].toBoolean()
```
rename to `labelsPreviousCommand`

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,23 @@
+            }
+            val position = parts[1].toIntOrNull() ?: throw IllegalArgumentException("Second part must be type Int")
+            val labelPrevCmd = parts[2].toBoolean()
+            return LabelingCommandInformation(parts[0], position, labelPrevCmd)
```
Another neat kotlin declarative-style feature:

```suggestion
            return LabelingCommandInformation(parts.first(), position, labelPrevCmd)
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/TexifyConfigurable.kt:
```diff
@@ -1,34 +1,42 @@
 /**
  * @author Ruben Schellekens, Sten Wessel
  */
 class TexifyConfigurable(private val settings: TexifySettings) : SearchableConfigurable {
```
```suggestion
class TexifyConfigurable(private val settings: TexifySettings) : SearchableConfigurable {

```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/TexifyConfigurable.kt:
```diff
@@ -1,34 +1,42 @@
     private lateinit var automaticSecondInlineMathSymbol: JCheckBox
     private lateinit var automaticUpDownBracket: JCheckBox
     private lateinit var automaticItemInItemize: JCheckBox
+    private lateinit var table: TexifyConfigurableLabelCommands
```
Rename to something more sensible, like `labelDefiningCommands` for example.

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/TexifySettings.kt:
```diff
@@ -24,9 +25,50 @@ class TexifySettings : PersistentStateComponent<TexifySettings> {
+    /**
+     * internal list which stores the commands data
+     */
+    private val _labelCommands: HashMap<String, LabelingCommandInformation> =
```
Please don't use member prefixing. I find it weird that you need to use a backing field anyway.. Maybe find a way to join these in some form.

```kotlin
    /**
     * Internal list that stores the label defining commands.
     */
    private val labelCommandsOrABetterName: HashMap<String, LabelingCommandInformation> = hashMapOf(
            "\\label" to LabelingCommandInformation("\\label", 1, true)
    )
```

Stainzl replied:
The backing field is needed to initialize the Map in a not empty state. With the other property I could verify that every key of the map has a leading backslash.

RubenSchellekens replied:
Then you probably need a function that verifies this instead of a field. The member-prefixing still holds: don't do it :)

Maybe also take a look at [`Lazy`](https://kotlinlang.org/docs/reference/delegated-properties.html#lazy)?

Stainzl replied:
Could you explain why you wont use backing fields? I, personaly, think this is one of the best parts about kotlin, because I only have one field containing data, but could represent it in several to other functions.
My prefered way would be to write the `get()` directly under the field, but you can't initialize a property without a backing field. The reason why I don't use a function was that I wanted to do it in a more kotlin idiomatic style without a getter function.

stenwessel replied:
@Stainzl @RubenSchellekens I think backing fields are sometimes necessary in Kotlin to provide some modification to the underlying data structure. However, in this particular case, I find it a bit odd. Especially since the getter performs a map on the keys on every access. This is an (undocumented) and unconventional side effect of the getter; the caller does not expect this calculation every time.

For me, this raises the question why the ```addLeadingSlash``` logic is placed here. I think this is not the responsibility of the ```TexifySettings``` class, but rather of the classes that directly manipulate ```labelCommands```.

Therefore, you could simply use ```val labelCommands = hashMapOf(...)``` directly and expose the mutable map. This is consistent with the other class members as everything is publicly modifiable already. Note in the docstring keys are required to have a leading slash. I think this is sufficient.

Side note: if you really really want to use a backing field here (which I think you shouldn't), do something like this (reducing the computations in the getter):
```kotlin
private val _labelCommands: HashMap<String, LabelingCommandInformation> = hashMapOf(...)
val labelCommands: Map<String, LabelingCommandInformation> get() =  _labelCommands

fun addLabelCommand(key: String, info: LabelingCommandInformation) {
    _labelCommands[key.addLeadingSlash()] = info
}
```

Stainzl replied:
Okay, with the use of a external state class the backing field couold be removed.

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/TexifySettings.kt:
```diff
@@ -24,9 +25,50 @@ class TexifySettings : PersistentStateComponent<TexifySettings> {
+     * Do not use this field anywhere! It is only needed to persist the data of the commands to the settings.
+     * It maps any command to its string representation and on first load back to the object
+     */
+    var labelCommandsAsString: Map<String, String>
```
It is not used? Why is it here then? This is probably not the correct way of doing this. I'd suggest handling the input in dedicated accessor functions (like `addCommand`?) to abstract the process.

Stainzl replied:
It is used by IntelliJ's function to store settings. IntelliJ could only store simple lists or maps with simple types. As I need to store a complex type this field maps the complex type to a string representation which could be stored by IntelliJ.
I hope you get what I wanted to do. If you have a better idea for this case, please let me hear it.

RubenSchellekens replied:
So, if I understand this correctly (and yes I've read the docs :P), this member's `set` gets called whenever the serialiser sets this property, thus automatically altering the state of the backing field?

This does feel like property abuse to be honest. I'd suggest just having a plain `var labelCommandsAsString` and modify it in `loadState` and `getState` to produce the serialised versions, so you separate state and behaviour a bit.

Stainzl replied:
Well, the `get` gets called every time the serialiser write the information to the `settings.xml` file and the `set` is called every time the deserialiser reads from the file.
I did it this way, because I don't want to store the same information twice. If there is only one field, it is in the same state every time, if there is mor than one field, you have to update both and when you don't to that, there is an inconsistency...
And there was already the backing field so it was the easiest way to reuse it and only remap the properties.
What do you mean exactly with "seperate state and behaviour"? The state is saved in the backing field and the behaviour is done by the diffrent properties calling to it.

stenwessel replied:
@Stainzl @RubenSchellekens There is actually a way to separate state and behavior here. See the first example in the docs: http://www.jetbrains.org/intellij/sdk/docs/basics/persisting_state_of_components.html?search=state#implementing-the-persistentstatecomponent-interface

Create a State class for serialization with the Map<String, String> object.
In the example, use a different type for `myState` to include the Map<String, LabelingCommandInformation> type, and handle the conversion in getState/loadState, where it semantically belongs. This avoids property abuse.

Note that since we only used 'simple' settings properties before it was not necessary to use a separate `State` class before. It will require some refactoring to migrate the other settings as well. Let me know if you need some help with that.

Stainzl replied:
Okay, I hope I understood it right. I tried it in commit 0e35e31a1ad4c4f5d00571372c8edc8f15ec1626. Could you review it and give me feedback?

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/TexifySettings.kt:
```diff
@@ -24,9 +25,50 @@ class TexifySettings : PersistentStateComponent<TexifySettings> {
+    val labelCommands: Map<String, LabelingCommandInformation>
+    get() {
+        return _labelCommands.mapKeys { it.key.addLeadingSlash() }
+    }
```
`get()` needs an indent.

Stainzl replied:
fixed in commit 4fc09f4f03cfda54b13164fb3b1d6e31c694f878

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/TexifySettings.kt:
```diff
@@ -24,9 +25,50 @@ class TexifySettings : PersistentStateComponent<TexifySettings> {
+    }
+
+    val labelAnyCommands: Map<String, LabelingCommandInformation>
+    get() = labelCommands.filter { it.value.labelPrevCmd }
```
`get()` needs an indent.

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+import javax.swing.text.PlainDocument
+
+/**
+ * window to enter the information about a label defining command
```
```suggestion
 * Window to enter the information about a label defining command.
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+ * window to enter the information about a label defining command
+ */
+class EditLabelDefiningCommand(cmdName: String, position: Int, labelAnyPrevCommandStatus: Boolean) :
+        DialogWrapper(null, false) {
```
Wrap Kotlin style. Also `cmdName` -> `commandName`

```kotlin
class EditLabelDefiningCommand(
        commandName: String,
        position: Int,
        labelAnyPreviousCommandStatus: Boolean
) : DialogWrapper(null, false) {
```

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+    private val commandName: JTextField = JTextField(cmdName, 25)
+    private val commandPosition: JTextField = JTextField("$position", 5)
+    private val labelAnyPrevCommand: JCheckBox = JCheckBox("Labels a previous command like \\section",
+            labelAnyPrevCommandStatus)
```
Split properly, also type is arguable unnecessary.
```kotlin
    private val labelAnyPrevCommand = JCheckBox(
            "Labels a previous command like \\section",
            labelAnyPrevCommandStatus
    )
```

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+class EditLabelDefiningCommand(cmdName: String, position: Int, labelAnyPrevCommandStatus: Boolean) :
+        DialogWrapper(null, false) {
+    private val commandName: JTextField = JTextField(cmdName, 25)
+    private val commandPosition: JTextField = JTextField("$position", 5)
```
```suggestion
    private val commandPosition = JTextField("$position", 5)
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+ */
+class EditLabelDefiningCommand(cmdName: String, position: Int, labelAnyPrevCommandStatus: Boolean) :
+        DialogWrapper(null, false) {
+    private val commandName: JTextField = JTextField(cmdName, 25)
```
```suggestion
    private val commandName = JTextField(cmdName, 25)
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+    override fun createCenterPanel(): JComponent? = null
+
+    /**
+     * create the layout of the window
```
```suggestion
     * Create the layout of the window.
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    /**
+     * validate the input so it is not empty and the Int value is in possible range
```
```suggestion
     * Validate the input so it is not empty and the Int value is in possible range.
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+
+    /**
+     * return the name of the command
+     */
```
either unnecessary, or wrongly documented

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+
+    /**
+     * return the position of the label
+     */
```
either unnecessary, or wrongly documented

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/EditLabelDefiningCommand.kt:
```diff
@@ -0,0 +1,104 @@
+
+    /**
+     * return whether the command labels any previous command or not
+     */
```
either unnecessary, or wrongly documented

Stainzl replied:
fixed in commit ed34f48cd6dd8da4a8ed3d61d6cf44599f4bc7dd

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/InputCommandFilter.kt:
```diff
@@ -0,0 +1,27 @@
+import javax.swing.text.AttributeSet
+import javax.swing.text.DocumentFilter
+
+class InputCommandFilter: DocumentFilter() {
```
```suggestion
class InputCommandFilter: DocumentFilter() {

```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/InputCommandFilter.kt:
```diff
@@ -0,0 +1,27 @@
+package nl.rubensten.texifyidea.settings.labeldefiningcommands
+
+import java.lang.StringBuilder
```
Unused import.

```suggestion
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/InputPossiblePositionFilter.kt:
```diff
@@ -0,0 +1,57 @@
+/**
+ * This class is used to prevent wrong input for the position of the label parameter
+ */
+class InputPossiblePositionFilter: DocumentFilter() {
```
```suggestion
class InputPossiblePositionFilter : DocumentFilter() {

```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/InputPossiblePositionFilter.kt:
```diff
@@ -0,0 +1,57 @@
+/**
+ * this function verifies that a String is a possible position fo the label parameter
+ */
+fun String.possiblePosition(): Boolean {
```
Don't bloat String. It's only intended for use in this class. Maybe consider putting it inside the class even.

```suggestion
private fun String.possiblePosition(): Boolean {
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/InputPossiblePositionFilter.kt:
```diff
@@ -0,0 +1,57 @@
+        (int >= 1)
+    } catch (ex : NumberFormatException) {
+        false
+    }
```
```kotlin
this.toIntOrNull()
```

Stainzl replied:
removed in commit 219692feebe4810461aa341700d8160051d2d9f7

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/LabelCommandSettingsTable.kt:
```diff
@@ -0,0 +1,27 @@
+class LabelCommandSettingsTable(model: TableModel, private var listener: ActionListener) : JBTable(model) {
+
+    /**
+     * create empty text and define a clickable link to add new command
```
```suggestion
     * Create empty text and define a clickable link to add a new command.
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/LabelCommandSettingsTable.kt:
```diff
@@ -0,0 +1,27 @@
+            }
+        }
+        test.text = "No command defined"
+        test.appendSecondaryText("Define new command", SimpleTextAttributes.LINK_ATTRIBUTES, listener)
```
@PHPirates: Input on this?
![image](https://user-images.githubusercontent.com/17410729/56140457-409b4d00-5f9b-11e9-8dab-fea1779fe233.png)

PHPirates replied:
I don't get that inspection. In fact, both in plugin.xml and in build.gradle the minimum version is 173.* so I wouldn't worry about it. I don't know where the 141 comes from, possibly an old module something still wandering around on your system.

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/LabelCommandSettingsTableModel.kt:
```diff
@@ -0,0 +1,13 @@
+/**
+ * tableModel to prevent editing the cell itself so the user is forced to use the edit window
+ */
+class LabelCommandSettingsTableModel : DefaultTableModel(), ItemRemovable {
```
```suggestion
class LabelCommandSettingsTableModel : DefaultTableModel(), ItemRemovable {

```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/LabelCommandSettingsTableModel.kt:
```diff
@@ -0,0 +1,13 @@
+import javax.swing.table.DefaultTableModel
+
+/**
+ * tableModel to prevent editing the cell itself so the user is forced to use the edit window
```
```suggestion
 * Table model to prevent editing the cell itself so the user is forced to use the edit window.
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/TexifyConfigurableLabelCommands.kt:
```diff
@@ -0,0 +1,231 @@
+/**
+ * Excluded table for the settings page to get smaller and more meaningful classes
+ */
+class TexifyConfigurableLabelCommands(private val settings: TexifySettings) {
```
```suggestion
class TexifyConfigurableLabelCommands(private val settings: TexifySettings) {

```

RubenSchellekens commented on src/nl/rubensten/texifyidea/settings/labeldefiningcommands/TexifyConfigurableLabelCommands.kt:
```diff
@@ -0,0 +1,231 @@
+ */
+class TexifyConfigurableLabelCommands(private val settings: TexifySettings) {
+    private var table: JBTable
+    private val tableInfo: LabelCommandSettingsTableModel = LabelCommandSettingsTableModel()
```
```suggestion
    private val tableInfo = LabelCommandSettingsTableModel()
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/structure/latex/LatexStructureViewElement.kt:
```diff
@@ -209,6 +206,16 @@ class LatexStructureViewElement(private val element: PsiElement) : StructureView
+        commands.filter { labelingCommands.containsKey(it.commandToken.text) }
+                .forEach {
+                    val element = LatexStructureViewCommandElement.newCommand(it)
+                            ?: return@forEach
```
can be oneliner

Stainzl replied:
fixed in commit 4fc09f4f03cfda54b13164fb3b1d6e31c694f878

RubenSchellekens commented on src/nl/rubensten/texifyidea/util/Labels.kt:
```diff
@@ -7,67 +7,84 @@ import nl.rubensten.texifyidea.index.BibtexIdIndex
-        .map { it.text.substringEnd(1) }
+//fun PsiFile.findBibtexLabelsInFileSet(): Sequence<String> = BibtexIdIndex.getIndexedIdsInFileSet(this)
+//        .asSequence()
+//        .map { it.text.substringEnd(1) }
```
Please remove commented out code.

Stainzl replied:
fixed in commit 4fc09f4f03cfda54b13164fb3b1d6e31c694f878

RubenSchellekens commented on src/nl/rubensten/texifyidea/util/Labels.kt:
```diff
@@ -7,67 +7,84 @@ import nl.rubensten.texifyidea.index.BibtexIdIndex
+        else {
+            true
         }
-        else true
```
```kotlin
        else true
```
is just fine ;)

RubenSchellekens commented on src/nl/rubensten/texifyidea/util/Labels.kt:
```diff
@@ -7,67 +7,84 @@ import nl.rubensten.texifyidea.index.BibtexIdIndex
+fun PsiFile.findLabelingCommands(): Collection<LatexCommands> = findLabelingCommandsSequence().toList()
+
+/**
+ * make a sequence of all commands which specify an label
```
```suggestion
 * Make a sequence of all commands that specify a label.
```

RubenSchellekens commented on src/nl/rubensten/texifyidea/util/Labels.kt:
```diff
@@ -76,13 +93,28 @@ fun PsiFile.findLabels(): Collection<PsiElement> {
+    return (bibtex + bibitem)
+}
+
+fun PsiFile.findBibitemCommands(): Sequence<LatexCommands> = LatexCommandsIndex.getItemsInFileSet(this).asSequence()
```
No documentation.

Stainzl replied:
fixed in commit e43885c396a6aec2c17f9d63fb39fa2ae1c7c550

RubenSchellekens commented on src/nl/rubensten/texifyidea/util/Magic.kt:
```diff
@@ -93,6 +93,11 @@ object Magic {
         )
 
+        /**
+         * LaTeX commands that increase an counter which could be labeled
```
```suggestion
         * LaTeX commands that increase a counter that can be labeled.
```

RubenSchellekens commented:
+resolve conflicts ;)

RubenSchellekens commented:
Any update on this?

Stainzl commented:
Sorry, I thought I wrote a comment already... I'm writing my exams in a week, so I can't get into the review right now. After my exams I will look at it.

RubenSchellekens commented:
Take your time! Good luck with your exams <3

PHPirates commented on src/nl/rubensten/texifyidea/util/Labels.kt:
```diff
@@ -7,67 +7,84 @@ import nl.rubensten.texifyidea.index.BibtexIdIndex
+fun PsiFile.findLabelingCommandsSequence(): Sequence<LatexCommands> {
+    val commandNames = TexifySettings.getInstance().labelCommands
+
+    return LatexCommandsIndex.getItemsInFileSet(this).asSequence()
```
Why is this getItemsInFile **Set** and not `getItems` like in the removed `PsiFile.findLabels()`? It seems to be causing the crash at #895, especially when I note the comment at https://github.com/Ruben-Sten/TeXiFy-IDEA/issues/454#issuecomment-374895880

RubenSchellekens replied:
`PsiFile.findLabels` should not be removed. Please add it back :)

PHPirates replied:
@RubenSchellekens On second thought, I do not agree. I think the new `findLabelingCommands()` _replaces_ `findLabels()`, because I think the whole point of that and this PR is to make sure the custom labeling commands are included - but @Stainzl  may correct me on this.

I believe the correct fix is replacing `getItemsInFileSet` with `getItems`, just as in the comment of that earlier crash report I mentioned.

You can reproduce the bug in this PR (just making sure there _is_ one...) by creating a file
`main.tex`:
```latex
\documentclass{article}
\begin{document}
    \input{included.tex}
\end{document}
```
`included.tex`:
```latex
\begin{figure}
    A figuring float.
    \label{fig:not-referenced}
\end{figure}
```
and you should (after maybe editing some files, you know the drill) get an exception like this one:
```
com.intellij.diagnostic.PluginException: Reported element LatexCommandsImpl(COMMANDS)[STUB]{\label} is not from the file 'LaTeX source file' the inspection 'LatexDuplicateLabel' (class nl.rubensten.texifyidea.inspections.latex.LatexDuplicateLabelInspection) was invoked for. Message: 'Duplicate label 'fig:not-referenced''.
```

I confirmed that the crashes stayed away when applying the suggested fix.

That this exception popped up for a _new_ inspection from #890 is basically a confusing coincidence I believe.

Stainzl replied:
Okay, I saw your comments in my mailbox and thougt I could answer quickly. I can confirm that my plan was to replace `findLabels()` completly and therefore I wrote a new function to check it, but don't break the whole thing. In the end I removed the old function. But as @RubenSchellekens pointed out in this [comment](https://github.com/Ruben-Sten/TeXiFy-IDEA/issues/895#issuecomment-492781202) it would be better to rename the new function back to the old name.

Stainzl replied:
I renamed the function back to the original name in commit 8e387e189139675d98b209ddcfc0198fc9186863

stenwessel commented on src/nl/rubensten/texifyidea/settings/TexifySettings.kt:
```diff
@@ -34,9 +34,50 @@ class TexifySettings : PersistentStateComponent<TexifySettings> {
     }
+
+    fun addCommand(cmd: LabelingCommandInformation) {
+        labelCommandsInternal[cmd.commandName] = cmd
```
Why is information from `LabelingCommandInformation` duplicated in the key? `commandName` now appears in the key as wel as the value. Maybe leave it out of `LabelingCommandInformation`, or is there a logical reason to keep it in there?

Stainzl replied:
The reason to have it in `LabelingCommandInformation` is, that I sometimes only have this object.
The map is used to get the searched element in a fast way. I thought about a list to store the single elements, but I decided for a map, because in a map a search for an element is much faster.
It could be possible to remove it, but in my opinion it is more logical to have all informations about an element in one class.
In the settings dialog I map the whole row to this class, without the field for the key I would have to return two values or something else.
I hope you understand what I wanted to say, why I decided to do it this way.

stenwessel commented on src/nl/rubensten/texifyidea/settings/TexifySettings.kt:
```diff
@@ -34,9 +34,50 @@ class TexifySettings : PersistentStateComponent<TexifySettings> {
+        labelCommandsInternal.remove(cmdName)
+    }
+
+    val labelAnyCommands: Map<String, LabelingCommandInformation>
```
I am not sure what this does? How does `AnyCommand` relate to `PreviousCommand`? Add clarification (or change the name)

Stainzl replied:
The commands in this list could label previous commands. I changed the name and added an comment for clarification.

stenwessel commented on src/nl/rubensten/texifyidea/settings/LabelingCommandInformation.kt:
```diff
@@ -0,0 +1,25 @@
+package nl.rubensten.texifyidea.settings
+
+data class LabelingCommandInformation(var commandName: String, var position: Int, var labelsPreviousCommand: Boolean) {
```
Add documentation of properties

stenwessel commented:
Great work! I do have so comments (mainly) about the settings class. Let me know if you agree or not :)

PHPirates commented:
In `Labels.kt` the functions `PsiFile.findLabels()` and `PsiFile.findLatexLabelsInFileSet()` are changed both here and in #782, which gives conflicts which I don't know how to resolve.

Stainzl commented:
Okay, I tried to merge #782 to this branch and pushed it to a new [branch](https://github.com/Stainzl/TeXiFy-IDEA/tree/add-labels-multi-include) in my fork.
But after the merge there is a really strange exception:
```
com.intellij.diagnostic.PluginException: Reported element LatexCommandsImpl(COMMANDS)[STUB]{\label} is not from the file 'LaTeX source file' the inspection 'LatexFigureNotReferenced' (class nl.rubensten.texifyidea.inspections.latex.LatexFigureNotReferencedInspection) was invoked for. Message: 'Figure is not referenced'.
Element' containing file: LaTeX source file
Inspection invoked for file: LaTeX source file
 [Plugin: nl.rubensten.texifyidea]
	at com.intellij.ide.plugins.PluginManagerCore.createPluginException(PluginManagerCore.java:368)
	at com.intellij.diagnostic.PluginProblemReporterImpl.createPluginExceptionByClass(PluginProblemReporterImpl.java:12)
	at com.intellij.diagnostic.PluginException.createByClass(PluginException.java:68)
	at com.intellij.diagnostic.PluginException.logPluginError(PluginException.java:84)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.createHighlightsForDescriptor(LocalInspectionsPass.java:545)
	at com.intellij.codeInsight.daemon.impl.LocalInspectionsPass.lambda$addDescriptorIncrementally$12(LocalInspectionsPass.java:427)
	at com.intellij.openapi.application.TransactionGuardImpl$2.run(TransactionGuardImpl.java:315)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.doRun(LaterInvocator.java:435)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.runNextEvent(LaterInvocator.java:419)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:403)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:747)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:696)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:391)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```
I tried it with the following example:
```tex
\documentclass[11pt]{article}
\usepackage{amsmath}
\usepackage{graphicx}
\begin{document}
    \begin{figure}[h]
        \includegraphics[width=\linewidth]{Broker.png}
        \caption{Bla}\label{blubber}
    \end{figure}
    \input{blub.tex, blubber.tex}
\end{document}
```
My structure looks like this:
```
src/
  blub.tex
  blubber.tex
  Broker.png
  main.tex
```
When you now execute the intention to create a the file a file with a leading space will be created, see this [comment](https://github.com/Ruben-Sten/TeXiFy-IDEA/pull/782#issuecomment-499790899). And when the file is created I get the exception.

Stainzl commented:
I noticed that in the structure view there are label entries with `#2` or other numbers, depending on which position the label is in the command. I tried to remove it there, but I don't know where I could say which labels should be listed there and which commands should not. Also the `missingLabelInspection` should be triggered when using such an label only declared in a `\newcommand` section I would say, but therefore I would also need a tip where I could do this...

PHPirates commented:
Please resolve merge conflicts (and why is this still not merged?), can't include it in the alpha build now

Stainzl commented:
This PR isn't merged since I'm waiting for a response [here](https://github.com/Hannah-Sten/TeXiFy-IDEA/pull/815#discussion_r291168165) and [here](https://github.com/Hannah-Sten/TeXiFy-IDEA/pull/815#discussion_r291165135). @HannahSchellekens and @stenwessel could you be so nice and answer that I could finish this PR?

HannahSchellekens commented:
@Stainzl is correct. And I'm also waiting for @stenwessel ;) If his issues are resolved, we're ready to merge!

At last!

Stainzl commented:
ping @stenwessel

Stainzl commented:
ping @stenwessel
